### PR TITLE
#2122 fix for 2 step mapping and generics

### PR DESCRIPTION
**Pull request by sjaakd**

### CODE REVIEW

**Comment by sjaakd** [comment]
It would be nice to have some more tricky tests.. E.g nested stuff:

`List<SimpleElement<T>>` ... or something with a `Map<String, T>` possibly with an implementation type delivered by the used method as result to test the new method in the factory full (`isAssignable`)...

@chris922 perhaps you see even more options.. I'm not to sure about the visitor.. I always find those tricky things..

**Comment by sjaakd** [comment]
and.. perhaps the method should not be part of type factory. but of type..

**Comment by sjaakd** [comment]
inheritance does not work yet..Just replace `Map` with `HashMap` in the target..  and also need to cleanup..

**Comment by sjaakd** [comment]
The test fails.. 

Somehow the wrong ref gets selected / returned on line 436 of the`MethodResolverImpl`. That is wrong and needs to be fixed. Probably deep in `MethodMatcher` (line 145)

However it also makes you think why the `resolveViaMethod` is called in the first place? 

We have the proper candidate in our hands, we just need to make a reference for it (methodY candidate) and even that we possibly should delay until we found a proper methodX.. (similar case for method-typeconversion). This is really a performance thing and needs to be addressed separately. 

I guess I'll make a separate issue out of that a s soon as I've got a solution.

**Comment by sjaakd** [comment]
probably the same problem occurs for other 2 steps:
* method, type conversion
* type conversion, method

we need to add test cases for this and add the solution also there. @chris922 : do you have some time to do this? I can help with the implementation (that's rather straightforward). Making test case will help.

**Comment by sjaakd** [comment]
Not sure why type conversion to method and vice versa was already working (in the master). Anyway, I guess its better to apply the same strategy everywhere. Added additional test as well.

File: processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java [hunk 1]
```diff
@@ -426,29 +426,33 @@ private Assignment resolveViaMethodAndMethod(Type sourceType, Type targetType) {
...
+                if ( Object.class.getName().equals( ySourceType.getName() ) ) {
                     //  java.lang.Object as intermediate result
                     continue;
                 }
 
-                methodRefY =
-                    resolveViaMethod( methodYCandidate.getSourceParameters().get( 0 ).getType(), targetType, true );
+                if ( ySourceType.isTypeVar() ) {
+                    ySourceType = ySourceType.resolveToType( targetType, methodYCandidate.getResultType() );
+                }
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java:437]: `resolveToType` already checks for `isTypeVar()`. You can directly do

```
ySourceType = ySourceType.resolveToType( targetType, methodYCandidate.getResultType() );
```

Comment (sjaakd) [reply processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java:437]: that's true..  but there's actually a reason for this:
1. I did not want to do the resolveToType unchecked in the `resolveToType` method itself.. Was a bit struggling with the method name.
2. I wanted to make the flow clear at the call site..

 I could have made it `resolveTypeVarToType` at hind site.  Will do this.

**Pull request merged**